### PR TITLE
[Wasm GC] Fuzz BrOn

### PR DESCRIPTION
**Pull request by kripken**
This found #6995

### CODE REVIEW

File: test/passes/translate-to-fuzz_all-features_metrics_noprint.txt [hunk 1]
```diff
@@ -1,53 +1,51 @@
...
- StringEq       : 3       
- StructNew      : 12      
+ [total]        : 718     
+ [vars]         : 30      
+ ArrayNew       : 20      
+ ArrayNewFixed  : 3       
+ AtomicFence    : 1       
+ Binary         : 77      
+ Block          : 79      
+ BrOn           : 1
```
Comment (kripken) [inline test/passes/translate-to-fuzz_all-features_metrics_noprint.txt:19]: got lucky here

File: src/tools/fuzzing/fuzzing.cpp [hunk 2]
```diff
@@ -3944,6 +3946,104 @@ Expression* TranslateToFuzzReader::makeRefCast(Type type) {
...
+      refType = Type(targetType.getHeapType(), getNullability());
+      break;
+    }
+    case BrOnCast: {
+      // The sent type is the heap type we cast to, with the input type's
+      // nullability, so the combination of the two must be a subtype of
+      // targetType.
+      castType = getSubType(targetType);
+      // The ref's type must be castable to the target, or we'd not validate.
+      refType = getSuperType(castType);
```
Comment (tlively) [inline src/tools/fuzzing/fuzzing.cpp:4019]: Maybe with some small probability we should have the `refType` be a subtype of the `castType`, which is valid and just makes the cast trivially succeed.

File: src/tools/fuzzing/fuzzing.cpp [hunk 3]
```diff
@@ -3944,6 +3946,104 @@ Expression* TranslateToFuzzReader::makeRefCast(Type type) {
...
+      if (targetType.isNonNullable()) {
+        // And it must have the right nullability for the target, as mentioned
+        // above.
+        refType = Type(refType.getHeapType(), NonNullable);
```
Comment (tlively) [inline src/tools/fuzzing/fuzzing.cpp:4023]: If the target type is non-nullable, either the `refType` or the `castType` must be non-nullable, but it's not necessary for them both to be non-nullable. We should be able to generate it all three ways.

File: src/tools/fuzzing/fuzzing.cpp [hunk 4]
```diff
@@ -3944,6 +3946,104 @@ Expression* TranslateToFuzzReader::makeRefCast(Type type) {
...
+      // The sent type is the ref's type, with adjusted nullability (if the cast
+      // allows nulls then no null can fail the cast, and what is sent is non-
+      // nullable).
+      refType = getSubType(targetType);
+      // The cast type is what flows out, which matters little here (as we fix
+      // up the flowing type anyhow), but the cast must be viable for
+      // validation, as with BrOnCast.
+      castType = getSubType(refType);
+      // There is no nullability to adjust: if targetType is non-nullable then
+      // both refType and castType are as well, as subtypes of it.
```
Comment (tlively) [inline src/tools/fuzzing/fuzzing.cpp:4037]: If the `targetType` is non-nullable, then `castType` should be allowed to be nullable.

File: test/passes/translate-to-fuzz_all-features_metrics_noprint.txt (hunk 1 above)
Comment (kripken) [reply test/passes/translate-to-fuzz_all-features_metrics_noprint.txt]: (easy come, easy go... update lost it)

File: src/tools/fuzzing/fuzzing.cpp [hunk 5]
```diff
@@ -1457,6 +1458,7 @@ Expression* TranslateToFuzzReader::_makenone() {
     .add(FeatureSet::GC | FeatureSet::ReferenceTypes, &Self::makeCallRef)
     .add(FeatureSet::GC | FeatureSet::ReferenceTypes, &Self::makeStructSet)
     .add(FeatureSet::GC | FeatureSet::ReferenceTypes, &Self::makeArraySet)
+    .add(FeatureSet::ReferenceTypes | FeatureSet::GC, &Self::makeBrOn)
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing/fuzzing.cpp:1461]: Good point. I fixed up a few more as well, all going with "references | GC", based on the logic of using the same order as the order of the proposals.

File: src/tools/fuzzing/fuzzing.cpp [hunk 6]
```diff
@@ -3944,6 +3946,104 @@ Expression* TranslateToFuzzReader::makeRefCast(Type type) {
...
+  // do that.
+  assert(targetType.isRef());
+  auto op = pick(BrOnNonNull, BrOnCast, BrOnCastFail);
+  Type castType = Type::none;
+  Type refType;
+  switch (op) {
+    case BrOnNonNull: {
+      // The sent type is the non-nullable version of the reference, so any ref
+      // of that type is ok, nullable or not.
+      refType = Type(targetType.getHeapType(), getNullability());
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing/fuzzing.cpp:4010]: Not sure I follow. That might conflict with the behavior of the other variants, as they all share the same make at the bottom?

It seems simpler to pick random nullability here, which lets `refType` be equal to the type that `make` will produce later.

File: src/tools/fuzzing/fuzzing.cpp (hunk 2 above)
Comment (kripken) [reply src/tools/fuzzing/fuzzing.cpp]: Good idea! Added.

File: src/tools/fuzzing/fuzzing.cpp (hunk 3 above)
Comment (kripken) [reply src/tools/fuzzing/fuzzing.cpp]: Good point, done.

File: src/tools/fuzzing/fuzzing.cpp (hunk 4 above)
Comment (kripken) [reply src/tools/fuzzing/fuzzing.cpp]: Good idea, done.

File: src/tools/fuzzing/fuzzing.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlively) [reply src/tools/fuzzing/fuzzing.cpp:4010]: The first thing `make` does is choose a random subtype of the requested type, though. Choosing a random subtype of the allowed type before calling `make` is therefore redundant. (And I guess actually this applies to where we now randomly pick a subtype when setting `refType` in the `BrOnCast` and `BrOnCastFail` cases as well.)

[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing/fuzzing.cpp:4010]: Fair point. Though I'm not sure I'd like to depend on the behavior of `make` always picking a subtype. E.g. maybe in the future we'll make it do that only rarely, or something like that.

[the remark this answers is not in the document]
Comment (tlively) [reply src/tools/fuzzing/fuzzing.cpp:4010]: Fair enough 👍

**Review by tlively — APPROVED** [review]

**Pull request merged**